src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -28928,6 +28928,244 @@ public void Init(T t)
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "F").WithLocation(44, 18));
+        }
+
+        [Fact, WorkItem(49964, "https://github.com/dotnet/roslyn/issues/49964")]
+        public void AdjustPropertyState_MaybeDefaultInput_MaybeNullOutput()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class C<T> {
+    [AllowNull]
+    T Prop { get; set; }
+
+    public C() { }
+
+    public C(int unused)
+    {
+        M1(Prop); // 1
+    }
+
+    public void M()
+    {
+        Prop = default;
+        M1(Prop); // 2
+        Prop.ToString(); // 3
+    }
+
+    public void M1(T t) { }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, AllowNullAttributeDefinition });
+            // It's not clear whether diagnostic 1 and 2 are appropriate.
+            // Going by the principle of "if you put a good value in, you will get a good value out",
+            // it seems like property state should be MaybeNull after assigning a MaybeDefault to the property.
+
+            // Note that the lack of constructor exit warning in 'C()' is expected due to presence of 'AllowNull' on the property declaration.
+            comp.VerifyDiagnostics(
+                // (12,12): warning CS8604: Possible null reference argument for parameter 't' in 'void C<T>.M1(T t)'.
+                //         M1(Prop); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "Prop").WithArguments("t", "void C<T>.M1(T t)").WithLocation(12, 12),
+                // (18,12): warning CS8604: Possible null reference argument for parameter 't' in 'void C<T>.M1(T t)'.
+                //         M1(Prop); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "Prop").WithArguments("t", "void C<T>.M1(T t)").WithLocation(18, 12),
+                // (19,9): warning CS8602: Dereference of a possibly null reference.
+                //         Prop.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "Prop").WithLocation(19, 9));
+        }
+
+        [Fact, WorkItem(49964, "https://github.com/dotnet/roslyn/issues/49964")]
+        public void AdjustPropertyState_Generic_MaybeNullInput_NotNullOutput()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class C<T> {
+    [NotNull]
+    T Prop { get; set; }
+
+    public C() // 1
+    {
+    }
+
+    public C(T t) // 2
+    {
+        Prop = t;
+    }
+
+    public void M(T t) {
+        Prop = t;
+        Prop.ToString(); // 3
+    }
+
+    public void M1(T t) { }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, NotNullAttributeDefinition });
+            // It's not clear whether diagnostic 2 or 3 are appropriate.
+            // Going by the principle of "if you put a good value in, you will get a good value out",
+            // it seems like property state should be NotNull after assigning a MaybeNull to the property.
```

Comment:
My mental model is that this sample is not meaningfully different than the following which does not warn:

```c#
#nullable enable

public class C {
    [return:NotNull]T M<T>(T t) => t;
}
```

Reply:
This appears to warn on [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABABguYEYA6AEQEsIAc1wB7WPwDGUbgGFRKOAEEahAJ5R+UANyMWAZg4AmCrIoBvRhWsUA2swDsIAHKiYzooQC6AFQoBZAB4fAD4ACj8YAEoKAF4QihhdBgBfIA==)

Reply:
I copied that from SharpLab where it was not warning ... I must have been stuck on an old branch. Is it Friday yet? 

Still though I do believe my comment about this conceptually being the same problem here. Hence I would expect them to both or neither warn.

Reply:
Looking closer: this test reveals an issue that this PR is not intended to solve.

We have a pattern in nullable for properties where setters are allowed to accept a "wider" type than the getter will return, on the assumption that the setter will "sanitize" the input and ensure that the getter always returns a value of the correct type.

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BYAUAYgDsBXAG1ImFLgAI5DLq8ABABhuYEYA6AEQEsIAc0IB7WPwDGUbgGFRKOAEEGpAJ5R+UANx48ANwgIaAdxoBeGoThmA6vxRC4MABQBKXbhPcAKnAAeMBZWZKSe3n6BvqIAyjAI/IRC7to0APRpVqKmRoSJQnT+AA5wkvAoerjMAMwcAEw09o7OeADeeDSdHJysAPw0APrwgZ5dNADaSuSiJgByoQC6HV013eyRMMud7bhjY05B5gB8g8NBvf0ARJeje51QzhYnQwGHNIakxHC3nQC+eL8gA===)
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

var w = new Widget();
w.Text = null;
w.Text.ToString(); // no warning expected

public class Widget
{
    string? _text;
    [AllowNull]
    public string Text
    {
        get => _text ?? "";
        set => _text = value;
    }
}
```

These tests raise, but do not answer, the question of whether this behavior should be extended from just the "maybe null/not null" difference to the "maybe default/maybe null" or "maybe null/`[NotNull] T`" scenarios.

The WorkItem attribute on these tests indicates this, however for clarity I'm happy to add a link to the follow-up issue inline.